chia/plotting/plot_tools.py

```diff
@@ -106,188 +106,376 @@ def stream_plot_info_ph(
+        for directory_name in self.get_plot_directories():
+            directory = Path(directory_name).resolve()
+            all_files[directory] = _get_filenames(directory)
+        return all_files
+
+    def plot_count(self):
+        with self:
+            return len(self.plots)
+
+    def add_plot_directory(self, str_path: str) -> Dict:
+        log.debug(f"add_plot_directory {str_path}")
+        config = load_config(self.root_path, "config.yaml")
+        if str(Path(str_path).resolve()) not in self.get_plot_directories(config):
+            config["harvester"]["plot_directories"].append(str(Path(str_path).resolve()))
+        save_config(self.root_path, "config.yaml", config)
+        self.trigger_refresh()
+        return config
+
+    def remove_plot_directory(self, str_path: str) -> None:
+        log.debug(f"remove_plot_directory {str_path}")
+        config = load_config(self.root_path, "config.yaml")
+        str_paths: List[str] = self.get_plot_directories(config)
+        # If path str matches exactly, remove
+        if str_path in str_paths:
+            str_paths.remove(str_path)
+
+        # If path matcehs full path, remove
+        new_paths = [Path(sp).resolve() for sp in str_paths]
+        if Path(str_path).resolve() in new_paths:
+            new_paths.remove(Path(str_path).resolve())
+
+        config["harvester"]["plot_directories"] = [str(np) for np in new_paths]
+        save_config(self.root_path, "config.yaml", config)
+        self.trigger_refresh()
+
+    def remove_plot(self, path: Path):
+        log.debug(f"remove_plot {str(path)}")
+        # Remove absolute and relative paths
+        if path.exists():
+            path.unlink()
+
+        self.trigger_refresh()
+
+    def needs_refresh(self) -> bool:
+        return time.time() - self.last_refresh_time > float(self.refresh_parameter.interval_seconds)
+
+    def start_refreshing(self):
+        self._refreshing_enabled = True
+        if self._refresh_thread is None or not self._refresh_thread.is_alive():
+            self._refresh_thread = threading.Thread(target=self._refresh_task)
+            self._refresh_thread.start()
+
+    def stop_refreshing(self):
+        self._refreshing_enabled = False
+        if self._refresh_thread is not None and self._refresh_thread.is_alive():
+            self._refresh_thread.join()
+            self._refresh_thread = None
+
+    def trigger_refresh(self):
+        log.debug("trigger_refresh")
+        self.last_refresh_time = 0
+
+    def _refresh_task(self):
+        while self._refreshing_enabled:
+
+            while not self.needs_refresh() and self._refreshing_enabled:
+                time.sleep(1)
+
+            total_result: PlotRefreshResult = PlotRefreshResult()
+            while self.needs_refresh() and self._refreshing_enabled:
+                batch_result: PlotRefreshResult = self.refresh_batch()
+                total_result += batch_result
+                if self._refresh_callback is not None:
```

Comment:
in what case would _refresh_callback be None ?

Reply:
Whenever `PlotManager` constructor gets created without `refresh_parameter` set like here https://github.com/Chia-Network/chia-blockchain/blob/5947039889bf16026d6fea9ed0b6af93209c98d3/chia/plotting/create_plots.py#L163 or here https://github.com/Chia-Network/chia-blockchain/blob/5947039889bf16026d6fea9ed0b6af93209c98d3/chia/cmds/plots.py#L27

Not that the refreshing gets started in those particular cases but there _might_ be others?

Reply:
Refactored, see cabfd5d and 88038ff